engine/consensus/matching/core.go

```diff
@@ -695,6 +699,71 @@ func (c *Core) sealableResults() ([]*flow.IncorporatedResult, nextUnsealedResult
 	return results, nextUnsealeds, nil
 }
 
+// resultHasMultipleReceipts implements an additional _temporary_ safety measure:
+// only consider incorporatedResult sealable if we have AT LEAST 2 receipts committing
+// to the respective result from _different_ ENs
+func (c *Core) resultHasMultipleReceipts(incorporatedResult *flow.IncorporatedResult) bool {
+	blockID := incorporatedResult.Result.BlockID // block the result pertains to
+	resultID := incorporatedResult.Result.ID()
+
+	// get all receipts that are known for the block
+	receipts, err := c.receiptsDB.ByBlockIDAllExecutionReceipts(blockID)
+	if err != nil {
+		log.Error().Err(err).
+			Hex("block_id", logging.ID(blockID)).
+			Msg("could not get receipts by block ID")
+		return false
+	}
+
+	// index receipts for given incorporatedResult by their executor
+	// in case there are multiple receipts from the same executor, we keep the last one
+	receiptByExecutor := make(map[flow.Identifier]*flow.ExecutionReceipt) //map: executorID -> receipt
+	for _, receipt := range receipts {
+		if resultID != receipt.ExecutionResult.ID() { // skip receipts that commit to _different_ results
+			continue
+		}
+		receiptByExecutor[receipt.ExecutorID] = receipt
+	}
+
+	// too few receipts
+	if len(receiptByExecutor) < 2 {
+		return false
+	}
+	// At this point, we have at least two receipts from different executors and should be able to
+	// generate a candidate seal. We log information about all receipts for the result as evidence.
+	// Note that this code is probably run many times (until the seal has been finalized).
+	// However, as soon as we have at least 2 receipts, the matching Engine will produce a candidate
+	// seal and put it in the mempool. To prevent spamming the logs, we log only when there is
+	// no seal in the IncorporatedResultSeals mempool:
+	_, sealAlreadyGenerated := c.seals.ByID(incorporatedResult.ID()) // IncorporatedResultSeal has the same ID as incorporatedResult
+	if !sealAlreadyGenerated {
+		header, err := c.headersDB.ByBlockID(blockID)
+		if err != nil {
+			log.Fatal().Err(err).
+				Hex("block_id", logging.ID(blockID)).
+				Msg("could not header for block")
+			return false
+		}
+
+		l := c.log.Info().
```

Comment:
would this generate lots of info log?

Reply:
Happy to change it. Let me know what you prefer. 
The logging should essentially be almost identical to the initial version you built: https://github.com/onflow/flow-go/pull/363#discussion_r569732038 . The only noteworthy differences between your version and the revised behaviour as of this PR:
* `FirstUnmatchedChunkIndex: -1` is now omitted.
* `SealedByEmergency` is only included when we actually evaluated the emergency sealing condition (as opposed to before, where the default value was false even if we didn't check whether the result qualified for emergency sealing)
* we also have information whether 2 receipts for the result are known

Reply:
I would suggest to use `debug` log. 

The log I built only print logs for next unsealed result, whereas the added log here will print log for all unsealed results. 

Since you've added extensive logging for the next unsealed result, I think this logging could be debug level